### ISSUE TITLE
"Error processing record: Input length = 1"

### ISSUE THREAD

**Issue by danielcbright** [codekitchen/logstash-input-kinesis#6]
I'm processing CloudWatch logs (via Kinesis subscription) and am getting the following error for all entries, no matter the length. Thoughts?

"Error processing record: Input length = 1", :level=>:error, :file=>"logstash/inputs/kinesis/worker.rb", :line=>"54", :method=>"process_record"

**Comment by codekitchen** [comment]
Hm I haven't tried processing a cloudwatch log subscription with this library yet, so I can only guess. Reading through http://docs.aws.amazon.com/AmazonCloudWatch/latest/DeveloperGuide/Subscriptions.html , it sounds like the log lines are base64 encoded (which is normal in kinesis) but then also gzipped.

I wonder if that compression is the problem? Right now the worker is hard-coded to always treat the event data as a UTF-8 string, see https://github.com/codekitchen/logstash-input-kinesis/blob/master/lib/logstash/inputs/kinesis/worker.rb#L20 , maybe that decoding is failing because the data is still gzip compressed.

We should probably make this more flexible, and not always hard code to UTF-8.

**Comment by danielcbright** [comment]
That's a good question about whether it's compressed or not, I'll have to
ask our AWS contact that question, I think when it gets to Kinesis via a
subscriber it isn't compressed anymore (raw JSON) but it's further
encapsulated in more JSON. I'll get back ASAP.

Thanks

On Sat, Oct 31, 2015 at 11:51 AM, Brian Palmer notifications@github.com
wrote:

> Hm I haven't tried processing a cloudwatch log subscription with this
> library yet, so I can only guess. Reading through
> http://docs.aws.amazon.com/AmazonCloudWatch/latest/DeveloperGuide/Subscriptions.html
> , it sounds like the log lines are base64 encoded (which is normal in
> kinesis) but then also gzipped.
> 
> I wonder if that compression is the problem? Right now the worker is
> hard-coded to always treat the event data as a UTF-8 string, see
> https://github.com/codekitchen/logstash-input-kinesis/blob/master/lib/logstash/inputs/kinesis/worker.rb#L20
> , maybe that decoding is failing because the data is still gzip compressed.
> 
> We should probably make this more flexible, and not always hard code to
> UTF-8.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/codekitchen/logstash-input-kinesis/issues/6#issuecomment-152744860
> .

## 
- Daniel **Bright  *|Sr. Systems Engineer - Architecture & Platforms (Grid)
- w:\* 770.226.2840 _e:_ daniel.bright@weather.com

**Comment by albertogg** [comment]
Hi @codekitchen, this is happening to me too... As you said, the data does come in base64 format and gzipped. For that reason I tried to use a custom codec to decode that format inside the input but I believe the error described is happening somewhere inside your gem and breaks before reaching the codec so I don't know what to do.

```
logstash_1 | {:timestamp=>"2015-11-09T21:31:43.839000+0000", :message=>"Error processing record: Input length = 1", :level=>:error}
```

edit: I'm a complete newbie on logstash... but I can help you debug if you need help!

**Comment by papile** [comment]
@codekitchen is right that the data coming from Kinesis through the Cloudwatch Logs connector is indeed gzipped. If anyone is in a bind like I was with this needing to read this data in .... 

replace `raw = @decoder.decode(record.getData).to_s`

with:

```
 is = com.fasterxml.jackson.databind.util::ByteBufferBackedInputStream.new(record.getData)
 gzip = Java::java.util.zip::GZIPInputStream.new(is)
 raw = gzip.to_io.read
```

Keep in mind this is a mega hack/monkey patch which will make this work ONLY with gzip data. I think the permanent solution would be to have in the configuration an optional kinesis decoder that would default to UTF-8, but can also be gzip or other formats if the need arises. 

A common use case for this plugin though would be taking cloudwatch logs from kinesis and putting it somewhere like elasticsearch so +1 on adding gzip functionality.

**Comment by codekitchen** [comment]
I toyed with this a bit today, hoping that we could get away with just using a gzip codec in the logstash config. But it looks like that wouldn't fully solve the problem, as cloudwatch also sends multiple log lines in one big json message. Example from AWS docs:

``` json
{
    "owner": "123456789012",
    "logGroup": "CloudTrail",
    "logStream": "123456789012_CloudTrail_us-east-1",
    "subscriptionFilters": [
        "RootAccess"
    ],
    "messageType": "DATA_MESSAGE",
    "logEvents": [
        {
            "id": "31953106606966983378809025079804211143289615424298221568",
            "timestamp": 1432826855000,
            "message": "{\"eventVersion\":\"1.03\",\"userIdentity\":{\"type\":\"Root\"}"
        },
        {
            "id": "31953106606966983378809025079804211143289615424298221569",
            "timestamp": 1432826855000,
            "message": "{\"eventVersion\":\"1.03\",\"userIdentity\":{\"type\":\"Root\"}"
        },
        {
            "id": "31953106606966983378809025079804211143289615424298221570",
            "timestamp": 1432826855000,
            "message": "{\"eventVersion\":\"1.03\",\"userIdentity\":{\"type\":\"Root\"}"
        }
    ]
}
```

I think our best option is to add a flag, maybe something like `source => "cloudwatch_logs"` and follow a different decoding codepath when that's set. I considered making an independent `cloudwatch_logs` codec, but I don't think it adds any value since it's not really usable outside this input plugin anyway.

**Comment by papile** [comment]
I was able to get this working well with logstash filter splitting on `logEvents`. So just allowing gzip encoding would allow one to make this work with Cloudwatch Logs. However the split filter on logstash behaves oddly see (https://github.com/logstash-plugins/logstash-filter-split/issues/5#issuecomment-158479148) so I also had to patch that as well.

It would be indeed easier if the plugin parsed the log events anyway so the custom source would be nice.
